### PR TITLE
[nvfp4_training] Add CuteDSL NVFP4 RHT kernels

### PR DESCRIPTION
**Pull request by tbqh**
Added a CuteDSL (`nvidia-cutlass-dsl`) backend for the NVFP4 training RHT amax and quantization kernels, as a faster alternative to the Triton kernels from #4369.

### Benchmarks

Run environment: NVIDIA GB200, PyTorch 2.12.0a0, Triton 3.7.0, nvidia-cutlass-dsl.

**Hadamard amax** (`cutedsl_rht_amax` vs `triton_rht_amax`):

| Model | Shape | M | N | cutedsl_us | triton_us | speedup |
|---|---|---:|---:|---:|---:|---:|
| Llama 3 8B  | hidden-state input | 2048 | 4096  | 11.36 | 17.11 | 1.51x |
| Llama 3 8B  | mlp.down input     | 2048 | 14336 | 20.96 | 27.55 | 1.31x |
| Llama 3 70B | hidden-state input | 2048 | 8192  | 16.17 | 21.55 | 1.33x |
| Llama 3 70B | mlp.down input     | 2048 | 28672 | 39.58 | 43.20 | 1.09x |

**Hadamard quantize row+col** (`cutedsl_rht_quantize_row_col` vs `triton_rht_quantize_row_col`, RTNE):

| Model | Shape | M | N | cutedsl_us | triton_us | speedup |
|---|---|---:|---:|---:|---:|---:|
| Llama 3 8B  | hidden-state input | 2048 | 4096  | 13.89 | 24.36 | 1.75x |
| Llama 3 8B  | mlp.down input     | 2048 | 14336 | 34.91 | 60.68 | 1.74x |
| Llama 3 70B | hidden-state input | 2048 | 8192  | 23.73 | 37.21 | 1.57x |
| Llama 3 70B | mlp.down input     | 2048 | 28672 | 67.94 | 113.74| 1.67x |

**2D weight quantize** (`cutedsl_weight_quantize_2d` vs `triton_weight_quantize_2d`, no RHT):

| Model | Weight | M (out) | N (in) | cutedsl_us | triton_us | speedup |
|---|---|---:|---:|---:|---:|---:|
| Llama 3 8B  | mlp.gate/up | 14336 | 4096  | 65.98  | 248.33 | 3.76x |
| Llama 3 8B  | mlp.down    | 4096  | 14336 | 66.29  | 248.53 | 3.75x |
| Llama 3 70B | mlp.gate/up | 28672 | 8192  | 239.56 | 947.16 | 3.95x |
| Llama 3 70B | mlp.down    | 8192  | 28672 | 239.96 | 947.28 | 3.95x |

### End-to-end training throughput

Llama-3-style 1.2B, single GB200, seq_len 2048, batch 8 x grad_accum 96 (768 effective),
`torch.compile` (default mode). Measured with a whole-step CUDA-graphed training loop. Benchmarked against TransformerEngine (TE).

| Config | Quantize backend | Tok/s | ms/step | Speedup | Peak mem |
|---|---|---:|---:|---:|---:|
| AO NVFP4 (this PR, CUTEDSL) | CuteDSL | 140,775 | 11,173 | **1.29x** | 49.4 GB |
| AO NVFP4 (existing, TRITON) | Triton | 109,391 | 14,378 | 1.00x | 49.4 GB |
| TE NVFP4 | TE fused | 145,999 | 10,773 | 1.33x | 54.1 GB |
| TE bf16 | — | 114,681 | 13,715 | 1.05x | 58.0 GB |

### Convergence Run

Code_enFUk0UTBg

### CODE REVIEW

File: torchao/prototype/moe_training/nvfp4_training/nvfp4_training.py [hunk 1]
```diff
@@ -78,10 +78,15 @@ class NVFP4TrainingConfig(AOBaseConfig):
     Args:
         kernel_preference: Backend for quantization kernels.
             TRITON: Pure-Triton RHT + stochastic rounding path.
+            CUTEDSL: CuteDSL kernels for the full quantize path (amax, forward
```
Comment (vkuzo) [inline torchao/prototype/moe_training/nvfp4_training/nvfp4_training.py:81]: I'd recommend separating the kernel implementation language (triton vs cute) vs the numerical recipe (in this case, 16x16 vs 1x16 weight inner block size).  Can we do something like:
1. add KernelPreference.CUTEDSL (already in this PR)
2. add a new flag for weight quantization block size 1x16 vs 16x16
3. throw an error if a user selects a combination of weight quant block size and kernel preference which does not have kernels

File: torchao/quantization/quantize_/common/kernel_preference.py [hunk 2]
```diff
@@ -48,5 +48,13 @@ class KernelPreference(str, Enum):
     Requires bfloat16 input; M, K, N all divisible by 128.
     """
 
+    CUTEDSL = "cutedsl"
+    """Use CuTeDSL kernels for the full NVFP4-training quantize path: RHT amax,
```
Comment (vkuzo) [inline torchao/quantization/quantize_/common/kernel_preference.py]: this enum is repo-level, but the docblock is currently for the nvfp4 training kernels.  Can we keep this docblock generic

**Comment by vkuzo** [comment]
@claude review.  Focus on: (a) do all the new cute kernels have tests for bitwise equivalency with the old triton kernels, (b) numerical correctness, (c) edge cases, (d) how well the existing repo structure was followed

**Comment by vkuzo** [comment]
claude review looks good, here's my take:
1. can we separate numeric recipe change (1x16 vs 16x16) from kernel selector
2. code structure wise, can we make the benchmarks and tests parametrize over the kernel selector instead of copy paste.  For example, IMO `benchmarks/prototype/nvfp4_training/bench_hadamard_amax_cutedsl.py` should instead be `bench_hadamard_amax.py` which loops over (triton, cutedsl). This way it's clear what is shared vs different.

after those are done should be good to go! Thanks for working on this, and the perf improvements are very nice!

**Comment by tbqh** [comment]
@vkuzo Regarding those points:

1. There should not have been any numeric recipe change, that was my mistake. I updated everything to make the CuteDSL kernels match the existing Triton 16x16 recipe. This is a medium-size change, and it might be worth looking over the PR again.
2. The tests and benchmarks are now parameterized over the quantization backends.

**Comment by vkuzo** [comment]
@claude review again

**Review by vkuzo — APPROVED** [review]
thanks for addressing the changes! could you rebase and also make sure `ruff` passes? thank you!
